mtags/src/main/scala-3/scala/meta/internal/pc/completions/MatchCaseCompletions.scala

```diff
@@ -0,0 +1,418 @@
+    if definitions.isTupleClass(selectorSym) || definitions.isFunctionClass(
+        selectorSym
+      )
+    then
+      val label =
+        if !patternOnly then s"case ${parents.selector.show} =>"
+        else parents.selector.show
+      result += CompletionValue.CaseKeyword(
+        selectorSym,
+        label,
+        Some(
+          if !patternOnly then
+            if config.isCompletionSnippetsEnabled() then "case ($0) =>"
+            else "case () =>"
+          else if config.isCompletionSnippetsEnabled() then "($0)"
+          else "()"
+        ),
+        Nil,
+        range = Some(completionPos.toEditRange),
+        command = config.parameterHintsCommand().asScala,
+      )
+    end if
+
+    val res = result.result()
+    res
+  end contribute
+
+  /**
+   * A `match` keyword completion to generate an exhaustive pattern match for sealed types.
+   *
+   * @param selector the match expression being deconstructed or `EmptyTree` when
+   *                 not in a match expression (for example `List(1).foreach { case@@ }`.
+   * @param completionPos the position of the completion
+   * @param typedtree typed tree of the file, used for generating auto imports
+   * @param indexedContext
+   * @param config
+   */
+  def matchContribute(
+      selector: Tree,
+      completionPos: CompletionPos,
+      typedTree: Tree,
+      indexedContext: IndexedContext,
+      config: PresentationCompilerConfig,
+  ): List[CompletionValue] =
+    import indexedContext.ctx
+    val pos = completionPos.sourcePos
+    val definitions = indexedContext.ctx.definitions
+    val text = pos.source.content().mkString
+    val clientSupportsSnippets = config.isCompletionSnippetsEnabled()
+    lazy val autoImportsGen = AutoImports.generator(
+      pos,
+      text,
+      typedTree,
+      indexedContext,
+      config,
+    )
+    val completionGenerator = CompletionValueGenerator(
+      indexedContext,
+      completionPos,
+      clientSupportsSnippets,
+    )
+    val result = ListBuffer.empty[CompletionValue]
+    val tpe = selector.tpe.widen.bounds.hi match
+      case tr @ TypeRef(_, _) => tr.underlying
+      case t => t
+
+    def subclassesForType(tpe: Type)(using Context): List[Symbol] =
+      tpe.typeSymbol.info
+      val subclasses = ListBuffer.empty[Symbol]
+      val parents = ListBuffer.empty[Symbol]
+      // Get parent types from refined type
+      def loop(tpe: Type): Unit =
+        tpe match
+          case AndType(tp1, tp2) =>
+            loop(tp1)
+            loop(tp2)
+          case t => parents += tpe.typeSymbol
+      loop(tpe.widen.bounds.hi)
+      parents.toList.foreach {
+        _.sealedStrictDescendants.foreach(sym =>
+          if !(sym.is(Sealed) && (sym.is(Abstract) || sym.is(Trait))) then
+            subclasses += sym
+        )
+      }
+      val subclassesResult = subclasses.toList
+      subclassesResult
+    end subclassesForType
+
+    val sortedSubclasses = subclassesForType(tpe)
```

Comment:
Can't we just get all directSubclasses by [sym.children](https://github.com/lampepfl/dotty/blob/e560c2d6970c864ceea5607597d74cf49c443965/compiler/src/dotty/tools/dotc/core/SymDenotations.scala#L1601-L1632) instead of traverse all the subclasses by `sealedStrictDescendents`?

I think what we're interested in here (exhaustive pattern match) is only it's direct subclasses 🤔

Reply:
`sealedStrictDescendants` is for sealed traits and sealed abstract classes, using it we can be more specific. Maybe we can use `children` and if a child is a `sealed` trait/abstract class do a loop or just use it without "splitting" it more?

Reply:
Ahh, never mind, I misunderstood the behavior of `sealedStrictDescendants`.